rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+to HTTP polling delivery mechanisms: passing information with a response header (as done for licensing), WebSockets, ServerSideEvents.
+
+#### Notification Status
+The delivery mechanism handles Notification status change performed by Kibana UI.
+Every notification has a different status for every user.
+Supported notification statuses:
+- `read: boolean`
+- `pinned: boolean`
+
+Kibana UI changes notification status to `read: true` when a user interacts with it: clicks on CTA or `Mark as Read` button.
+
+The user can change the notification status to `read: false` manually in the Kibana UI.
+
+The user can change the status for all the existing notifications (`mark all as read`), even for those not-rendered in UI yet.
+
+Kibana HTTP API:
+- mark notification as read
+```
+POST /notification/read
+{ notification_id: string[] }
+```
+- Mark all the notification read
+```
+POST /notifications/_mark_all_read
+```
+- Pin a notification.
+```
+POST /notification/pin
+{ notification_id: string[] }
+```
+- Unpin a notification.
+```
+POST /notification/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Threat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
```

Comment:
A few of other benefits of this approach come to mind for me:

#### Graceful degradation

If the UNS is unavailable for any reason (even a planned/scheduled outage), then all cloud hosted instances will be unable to publish or retrieve notifications, even if these notifications were generated by that very Kibana instance.

If we took this pluggable storage approach, then locally sourced notifications could still work, as they wouldn't rely on the UNS for storage or retrieval.

To me, this is similar to your Netflix feed: the different categories are managed by different services to provide a single UX. If one of those services is unavailable, then the rest of your feed continues to work.

#### Extensibility

Once solutions start taking advantage of the notification service, we might uncover new requirements, which may in turn require changes to the data model. If we are coupled to the UNS's representation of the data model, then we are potentially limited by what we can do (short of a `_meta` dumping ground).

If, on the other hand, we store locally sourced events locally, then we have full control over the data model, and we can perform data migrations as needed.

We would still need to support the UNS's data model, but the types of events created by external sources may not require the same amount of functionality as locally sourced events.

#### Flexible authorization model

We could choose to authorize locally sourced notifications differently from externally sourced notifications. The UNS (i.e. externally sourced notification) is likely in a better position to perform auth on their own, but locally sourced notifications could have a stack-specific ACL attached.


#### Migration to ESS?

If an on-prem user wants to move their cluster to the Cloud, then a snapshot/restore would allow them to retain their existing notifications and notification history.

Reply:
Having thought about this some more, I don't think many of these benefits really apply if locally sourced notifications have to be available consistently in other Elastic products

Reply:
> Graceful degradation
If the UNS is unavailable for any reason (even a planned/scheduled outage), then all cloud hosted instances will be unable to publish or retrieve notifications, even if these notifications were generated by that very Kibana instance.

Later, when UNS is available, it should synchronize notifications and their state between the remote and local storage. For sure, it's undoable, but it's not simple.

> Extensibility
Once solutions start taking advantage of the notification service, we might uncover new requirements, which may in turn require changes to the data model. 

Agree. It might be useful if we are ready to put up with the cost of data and UI inconsistency between a notification shown in Kibana UI and Console UI.

> Flexible authorization model
We could choose to authorize locally sourced notifications differently from externally sourced notifications. The UNS (i.e. externally sourced notification) is likely in a better position to perform auth on their own, but locally sourced notifications could have a stack-specific ACL attached.

Based on the requirements that Cloud must show Kibana notifications in Cloud Console UI, how would it implement the Kibana-compatible authorization model? 

> Migration to ESS?
If an on-prem user wants to move their cluster to the Cloud, then a snapshot/restore would allow them to retain their existing notifications and notification history.

That's a valid point as long as we are sure that a user migrates to deployment with the same configuration as the local one.